### PR TITLE
PARQUET-2448: Fix model conversions for avro <= 1.8 generated classes

### PR DESCRIPTION
**Pull request by RustedBones**
When running with avro <= 1.8, the default data supplier was looking for the root record's class conversions to initialize its model.

In case the record contained a nested class with another logical-type, the conversion will fail
Adapt the code to inspect the whole schema and initialize the model accordingly.

### CODE REVIEW

**Review by clairemcginty — COMMENTED** [review]
Great catch @RustedBones . This change looks good to me  🚢

**Comment by wgtmac** [comment]
Thanks for the fix! Could you please make the CI happy?

**Comment by wgtmac** [comment]
I need to spend some time to get familiar with the code path.

cc avro experts @gszadovszky @Fokko .

File: doc/dremel_paper/dremel_example.png
Comment (gszadovszky) [inline doc/dremel_paper/dremel_example.png:1]: Something went odd with the formatting. It shouldn't touch binary files.

File: doc/dremel_paper/schema.png
Comment (gszadovszky) [inline doc/dremel_paper/schema.png:1]: Same as above

File: parquet-avro/src/test/java/org/apache/parquet/avro/TestAvroRecordConverter.java
Comment (gszadovszky) [inline parquet-avro/src/test/java/org/apache/parquet/avro/TestAvroRecordConverter.java:1]: It seems you have rewritten the whole unit test. Was it necessary to do so? I would feel more comfortable if the existing tests would pass as they were.

File: parquet-avro/src/main/java/org/apache/parquet/avro/AvroRecordConverter.java [hunk 1]
```diff
@@ -215,31 +263,7 @@ static SpecificData getModelForSchema(Schema schema) {
...
-      final Conversion<?>[] conversions;
-      try {
-        conversionsField.setAccessible(true);
-        conversions = (Conversion<?>[]) conversionsField.get(null);
-      } catch (IllegalAccessException e) {
-        LOG.warn(String.format(
-            "Field `conversions` in class %s was inaccessible. Parquet will use default "
-                + "SpecificData model for reading and writing.",
-            clazz));
-        return null;
```
Comment (gszadovszky) [inline parquet-avro/src/main/java/org/apache/parquet/avro/AvroRecordConverter.java:235]: At this point we were returning a `null` as the model. The new code does not seem to do so. Is it intentional and correct?

**Review by gszadovszky — CHANGES_REQUESTED** [review]
I am not that experienced in the avro binding either but have some comments anyway. :)

File: parquet-avro/src/main/java/org/apache/parquet/avro/AvroRecordConverter.java (hunk 1 above)
Comment (RustedBones) [reply parquet-avro/src/main/java/org/apache/parquet/avro/AvroRecordConverter.java:235]: Good catch. Update the PR so the previous behavior is kept

File: parquet-avro/src/test/java/org/apache/parquet/avro/TestAvroRecordConverter.java
Comment (RustedBones) [reply parquet-avro/src/test/java/org/apache/parquet/avro/TestAvroRecordConverter.java:1]: Previous tests did not covered nested classes. I started to adapt the original code but that lead creating many classes with code duplication.
This refactoring went with the following simplification:
- test valid avro with generated class
- test deprecated avro (1.7 and 1.8 with same pseudo generated code)

Comment (gszadovszky) [reply parquet-avro/src/test/java/org/apache/parquet/avro/TestAvroRecordConverter.java:1]: OK, I accept this.
To be on the safe side, could you revert the unit test changes on your env and execute it on the new code?

**Review by gszadovszky — APPROVED** [review]

File: parquet-avro/src/test/java/org/apache/parquet/avro/TestAvroRecordConverter.java
Comment (RustedBones) [reply parquet-avro/src/test/java/org/apache/parquet/avro/TestAvroRecordConverter.java:1]: I can confirm that the original tests are passing with the changes

Comment (gszadovszky) [reply parquet-avro/src/test/java/org/apache/parquet/avro/TestAvroRecordConverter.java:1]: Thank you @RustedBones!

**Pull request merged**